### PR TITLE
feat: check if the upstream is using by route when deleting

### PR DESCRIPTION
**Pull request by starsz**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### New feature or improvement
- Describe the details and related test reports.

Now, when deleting an upstream, we don't check whether there is a route using it.
This may lead to the DP error if the user deletes the upstream by mistake.

So, we should check if the upstream is using by route when deleting.
If the upstream is using, we return 400 to guarantee the DP works well.

___
### Please add the corresponding test cases if necessary.

Added.
___

### CODE REVIEW

**Comment by nic-chen** [comment]
ci failed @starsz

**Review by membphis — APPROVED** [review]

**Review by nic-chen — COMMENTED** [review]
I think we also need to check wether used by service

**Comment by starsz** [comment]
> I think we also need to check wether used by service

Added.

File: api/test/e2enew/upstream/upstream_test.go [hunk 1]
```diff
@@ -723,3 +724,126 @@ var _ = ginkgo.Describe("Upstream update use patch method", func() {
...
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("delete upstream success", base.HttpTestCase{
+			Desc:         "delete upstream success",
+			Object:       base.ManagerApiExpect(),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/upstreams/u1",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+		}))
```
Comment (Jaycean) [inline api/test/e2enew/upstream/upstream_test.go:848]: Do we also need to confirm whether the upstream and service data are deleted？

Comment (starsz) [reply api/test/e2enew/upstream/upstream_test.go:880]: Yes, added.

File: api/test/e2enew/upstream/upstream_test.go [hunk 2]
```diff
@@ -826,3 +827,158 @@ var _ = ginkgo.Describe("Upstream update use patch method", func() {
 		})
 	})
 })
+
+var _ = ginkgo.Describe("test upstream delete (route is using)", func() {
```
Comment (tokers) [inline api/test/e2enew/upstream/upstream_test.go:831]: The description is not so good, `route is using` could be `route is in use`.

Comment (starsz) [reply api/test/e2enew/upstream/upstream_test.go:831]: Good advice. Let me fix it.

Comment (starsz) [reply api/test/e2enew/upstream/upstream_test.go:831]: Fixed.

**Review by tokers — APPROVED** [review]

**Review by nic-chen — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Review by imjoey — APPROVED** [review]
/lgtm

File: api/test/e2enew/upstream/upstream_test.go [hunk 3]
```diff
@@ -826,3 +827,158 @@ var _ = ginkgo.Describe("Upstream update use patch method", func() {
...
+			base.RunTestCase(tc)
+		},
+		table.Entry("create upstream without plugin", base.HttpTestCase{
+			Desc:    "create upstream without plugin",
+			Object:  base.ManagerApiExpect(),
+			Method:  http.MethodPut,
+			Path:    "/apisix/admin/upstreams/u1",
+			Headers: map[string]string{"Authorization": base.GetToken()},
+			Body: `{
+					"nodes": {
```
Comment (RedemptionC) [inline api/test/e2enew/upstream/upstream_test.go:843]: why isn't here a name field?

**Pull request merged**